### PR TITLE
Added a lint check for trait placement on hardcoded actor names

### PR DESCRIPTION
**Pull request by Mailaender**
Traits that belong on `Player`, `EditorPlayer`, `World` and `EditorWorld` can be placed freely which causes crashes or subtle errors that are hard to debug, but easy to check statically with the utility.

### CODE REVIEW

**Comment by pchote** [comment]
Nice. I'm pretty sure we have an issue associated with this, so if someone can find it we should mark this as closing it.

Can we please use an enum for the hardcoded actor types instead of strings? This would allow the compiler to guarantee that there aren't any typos.

File: OpenRA.Game/Traits/DebugPauseState.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@
 namespace OpenRA.Traits
 {
 	[Desc("Checks for pause related desyncs. Attach this to the world actor.")]
+	[TraitLocation("World")]
```
Comment (pchote) [inline OpenRA.Game/Traits/DebugPauseState.cs]: Style nit: IMO it is cleaner and clearer to keep the `[Desc()]` as the last attribute before the thing its attached to. When going through these to replace strings with the enum, can we also please move these above the descs?

Comment (penev92) [reply OpenRA.Game/Traits/DebugPauseState.cs]: As a general rule, definitions (variable and attribute) are best ordered by length. Anything else can be disruptive  when you want to be able to read a piece of code at a glance.

Comment (pchote) [reply OpenRA.Game/Traits/DebugPauseState.cs]: Indeed. This is one of the main things behind “cleaner and clearer” (descs are always going to be the longest attribute).

**Comment by Mailaender** [comment]
Added an `enum`, reordered the attribute, also tagged a bunch of more traits.

File: OpenRA.Game/Traits/LintAttributes.cs [hunk 2]
```diff
@@ -94,4 +94,14 @@ public PaletteReferenceAttribute(string playerPaletteReferenceSwitch)
 			PlayerPaletteReferenceSwitch = playerPaletteReferenceSwitch;
 		}
 	}
+
+	[AttributeUsage(AttributeTargets.Class)]
+	public sealed class TraitLocationAttribute : Attribute
+	{
+		public readonly SystemActors[] SystemActors;
```
Comment (pchote) [inline OpenRA.Game/Traits/LintAttributes.cs]: Switching to enums also removed the need to use an array here. Define `SystemActors` with `[Flags]` and give the members power-of-two values, then values can be composed using normal bitwise operations. Search for `[Flags]` and follow references to find examples of the many other places where we do this.

Comment (Mailaender) [reply OpenRA.Game/Traits/LintAttributes.cs]: Redid it with flagged enums.

File: OpenRA.Mods.Common/Lint/CheckTraitLocation.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+			foreach (var actorInfo in rules.Actors)
+			{
+				foreach (var traitInfo in actorInfo.Value.TraitInfos<TraitInfo>())
+				{
+					var traitLocation = traitInfo.GetType().GetCustomAttributes<TraitLocationAttribute>(true).FirstOrDefault();
+					if (traitLocation == null)
+						continue;
+
+					if (!Enum.TryParse(actorInfo.Key, true, out SystemActors systemActor))
+						continue;
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckTraitLocation.cs]: This prevents the check from working when a restricted trait is defined on any general actor (e.g. `1tnk`). We should be emitting an error here instead!

**Comment by Smittytron** [comment]
Needs rebase

**Comment by Mailaender** [comment]
Rebased.

**Comment by pchote** [comment]
Looks like you missed an important detail when rebasing that turned into a lint failure: `PlayerResources` is now required on the editorplayer so the editor resource value counter can function.

**Comment by Mailaender** [comment]
Sorry, forgot to run things locally as I didn't expect anything to change.

File: OpenRA.Game/Graphics/CursorProvider.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ public CursorProvider(ModData modData)
 
 			// Overwrite previous definitions if there are duplicates
 			var pals = new Dictionary<string, IProvidesCursorPaletteInfo>();
-			foreach (var p in modData.DefaultRules.Actors["world"].TraitInfos<IProvidesCursorPaletteInfo>())
+			foreach (var p in modData.DefaultRules.Actors[SystemActors.World.ToString().ToLowerInvariant()].TraitInfos<IProvidesCursorPaletteInfo>())
```
Comment (pchote) [inline OpenRA.Game/Graphics/CursorProvider.cs]: This is worse than hardcoding "world", IMO. Does anybody else have any thoughts about this?

Comment (pchote) [reply OpenRA.Game/Graphics/CursorProvider.cs]: ~~Something like `modData.DefaultRules.SystemActorInfo(SystemActors.World).TraitInfos<>()` that uses a switch statement internally might be a better compromise.~~

Comment (penev92) [reply OpenRA.Game/Graphics/CursorProvider.cs]: I don't agree this is worse. I would have gone with strings constants just for cleaner syntax, but this is still better than hardcoding "world".
Delegating the work to a centralized spot sounds good. Maybe even define the "special" actor names at the mod level - either in the manifest or in the rules, so they can be accessed similarly to what was suggested without requiring the mod to call them "world" and "player".

Comment (pchote) [reply OpenRA.Game/Graphics/CursorProvider.cs]: An even cleaner solution would be to change `modData.DefaultRules.Actors` to a new `ActorInfoDictionary` type that provides a custom indexer keyed on `SystemActors` in addition to strings. This would allow these lines to be written as `modData.DefaultRules.Actors[SystemActors.World].TraitInfos<>()` directly without requiring callers to use different methods for normal vs system actors.

You can copy/paste the old `ReadonlyDictionary` class https://github.com/OpenRA/OpenRA/blob/89854b71d1d48eece087f43ba37348eb57098c59/OpenRA.Game/Primitives/ReadOnlyDictionary.cs#L48-L98 which gives you a template of the overall structure - just change the generic types to `<string, ActorInfo>`, update the code style to modern OpenRA standards, and add the extra indexer.

**Comment by Mailaender** [comment]
Rebased and updated.

File: OpenRA.Game/Primitives/ActorInfoDictionary.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+using System.Collections;
+using System.Collections.Generic;
+
+namespace OpenRA
+{
+	public class ActorInfoDictionary : IReadOnlyDictionary<string, ActorInfo>
+	{
+		readonly Dictionary<string, ActorInfo> dict;
+
+		public ActorInfoDictionary()
```
Comment (pchote) [inline OpenRA.Game/Primitives/ActorInfoDictionary.cs]: This isn't used so can be removed. Or am I missing something?

File: OpenRA.Game/Primitives/ActorInfoDictionary.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		}
+
+		public bool ContainsKey(string key) => dict.ContainsKey(key);
+
+		public bool TryGetValue(string key, out ActorInfo value) => dict.TryGetValue(key, out value);
+
+		public int Count => dict.Count;
+
+		public ActorInfo this[string key] => dict[key];
+		public ActorInfo this[SystemActors key] => dict[key.ToString().ToLowerInvariant()];
```
Comment (pchote) [inline OpenRA.Game/Primitives/ActorInfoDictionary.cs:37]: This will incurr some extra allocations compared to a switch, but since anything on a hot path should be caching this result this approach does make sense from a maintainability perspective.

File: OpenRA.Mods.Common/Lint/CheckTraitLocation.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+					var locations = traitLocation.SystemActors.ToString();
+
+					if (!Enum.TryParse(actorInfo.Key, true, out SystemActors systemActor))
+					{
+						emitError("{0} does not belong on {1}. It is a system trait meant for {2}.".F(traitName, actorInfo.Key, locations));
+						continue;
+					}
+
+					if (!traitLocation.SystemActors.HasFlag(systemActor))
+						emitError("{0} does not belong on {1}. Put it on {2} instead.".F(traitName, actorInfo.Key, locations));
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckTraitLocation.cs]:
```suggestion
					if (!Enum.TryParse(actorInfo.Key, true, out SystemActors systemActor) || !traitLocation.SystemActors.HasFlag(systemActor))
					{
						// Remove the "Info" suffix
						var traitName = traitInfo.GetType().Name;
						traitName = traitName.Remove(traitName.Length - 4);

						var locations = traitLocation.SystemActors.ToString().Replace(", ", " or ");
						emitError("{0} does not belong on {1}. It is a system trait meant for {2}.".F(traitName, actorInfo.Key, locations));
					}
```

File: OpenRA.Mods.Common/Widgets/Logic/Editor/LayerSelectorLogic.cs [hunk 8]
```diff
@@ -36,10 +36,10 @@ public LayerSelectorLogic(Widget widget, WorldRenderer worldRenderer)
 			layerTemplateList.Layout = new GridLayout(layerTemplateList);
 			layerPreviewTemplate = layerTemplateList.Get<ScrollItemWidget>("LAYERPREVIEW_TEMPLATE");
 
-			IntializeLayerPreview(widget);
+			IntializeLayerPreview();
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/LayerSelectorLogic.cs]: This doesn't appear to be related to this PR?

**Review by pchote — COMMENTED** [review]
Looks good overall now. Just a couple of easily fixed comments.

Once this is merged we should update the `--docs` utility command to add "Requires system actor: X" the same way we currently add "Requires trait: X". All the "Attach to X" manually written in the trait descriptions can then be removed.

**Comment by pchote** [comment]
This PR interacts with several others (which add new traits or trait queries from the player/world actors) so i'd like to get this merged ASAP. I hope you don't mind that I pushed fixes directly to the branch to hopefully speed things up.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e17ef3014986b225f3f42be8798223697fada565...f3ab714041b9c5cfd1a767aed570ad1893f9367a
